### PR TITLE
SNOW-160505 Async Support

### PR DESCRIPTION
**Pull request by sfc-gh-jbahk**
### Description
This PR allows asynchronous querying by enabling the AsyncExec parameter to be set via the driver, which, when sent to the server, will not block waiting for the entire query to finish.

### Checklist
- [x] Code compiles correctly
- [x] Run ``make fmt`` to fix inconsistent formats
- [x] Run ``make lint`` to get lint errors and fix all of them
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: connection.go [hunk 1]
```diff
@@ -26,6 +26,12 @@ const (
 	statementTypeIDDelete           = statementTypeIDDml + int64(0x300)
 	statementTypeIDMerge            = statementTypeIDDml + int64(0x400)
 	statementTypeIDMultiTableInsert = statementTypeIDDml + int64(0x500)
+
+	statementTypeIDDdl     = int64(0x6000)
+	statementTypeIDCreate  = statementTypeIDDdl + int64(0x101)
+	statementTypeIDComment = statementTypeIDDdl + int64(0x200)
+	statementTypeIDDrop    = statementTypeIDDdl + int64(0x300)
+	statementTypeIDAlter   = statementTypeIDDdl + int64(0x401)
```
Comment (sfc-gh-stakeda) [inline connection.go:34]: why you need these?

**Comment by sfc-gh-stakeda** [comment]
@sfc-gh-jbahk , please add more comments to the PR or the code itself to explain what change you are making to help the reviewers get context quickly.

File: connection.go (hunk 1 above)
Comment (sfc-gh-jbahk) [reply connection.go:34]: It was requested in an open issue.

Comment (sfc-gh-stakeda) [reply connection.go:34]: It would be the best if you add this when it is required.

File: restful.go [hunk 2]
```diff
@@ -199,7 +203,31 @@ func postRestfulQueryHelper(
 
 		var resultURL string
 		isSessionRenewed := false
+		noResult, _ := isAsyncMode(ctx)
```
Comment (sfc-gh-cshi) [inline restful.go:206]: Should Restful as a wrapper on RESTFUL req/resp (transport layer, provide some hooks - 'Func'  at the most) be aware of that it has to write things to Rows or Result or even Rows/Result data structures that are specific to golang's sql module? Fill and return a wrapper object (e.g. QueryHandler) to include queryID, status, and statusChannel should be enough.  I think the original code that aimed to return queryId  only in async manner is simpler and more flexible.

Comment (sfc-gh-jbahk) [reply restful.go:206]: As discussed in the call today, the issue with returning a query handler object is that the user has no way to retrieve the `rows` or `result` object after. We can build a way for that to be possible, then our initial method of hanging on calls like `rows.Next()` is deemed moot, since that call depends on the async retrieval being completed. If we return a query handler and use the query ID to retrieve the result, the retrieval process will be kicked off only when the customer queries again. This method is more inline with existing code and APIs

Comment (sfc-gh-cshi) [reply restful.go:206]: I thought I already answered the question on Rows.Next() and we agreed to hanging on Rows.Next() is the simple way to code.  What was the concern raised on blocking at Rows.Next()? I think we agree that to make ExcuteContext and QueryContext async is the goal, and this PR is to allow first query to return immediately, but if future callers want to get queryId for this query they have to wait for the queryId channel to return something.

File: rows.go [hunk 3]
```diff
@@ -197,6 +213,14 @@ func (rows *snowflakeRows) NextResultSet() error {
 	return rows.ChunkDownloader.nextResultSet()
 }
 
+func (rows *snowflakeRows) checkAsyncQueryStatus() {
+	// if async query, block until query is finished
+	if rows.status == QueryStatusInProgress {
+		rows.status = <-rows.statusChannel
+		close(rows.statusChannel)
```
Comment (sfc-gh-anezvigin) [inline rows.go:220]: In general it's a good idea to close a channel on the writer side, not the reader side. Reading from a closed channel gives you a zero-value but writing to a closed channel panics. It's easier to spot a write on a closed channel if you do the closing around the place where you do the writing.

File: rows.go [hunk 4]
```diff
@@ -197,6 +213,14 @@ func (rows *snowflakeRows) NextResultSet() error {
 	return rows.ChunkDownloader.nextResultSet()
 }
 
+func (rows *snowflakeRows) checkAsyncQueryStatus() {
```
Comment (sfc-gh-anezvigin) [inline rows.go:216]: I would call this something like `waitForQueryStatusIfNeeded()` to make it clear at the call site that this function might block.

File: result.go [hunk 5]
```diff
@@ -1,26 +1,58 @@
...
-func (res *snowflakeResult) QueryID() string {
+func (res *snowflakeResult) GetQueryID() string {
 	return res.queryID
 }
+
+func (res *snowflakeResult) GetStatus() queryStatus {
+	return res.status
+}
+
+func (res *snowflakeResult) checkAsyncExecStatus() {
```
Comment (sfc-gh-anezvigin) [inline result.go:52]: Same comments as in rows.go

File: restful.go [hunk 6]
```diff
@@ -199,7 +203,31 @@ func postRestfulQueryHelper(
...
+				res.queryID = respd.Data.QueryID
+				res.status = QueryStatusInProgress
+				res.statusChannel = make(chan queryStatus)
+				respd.Data.AsyncResult = res
+			case "query":
+				rows.queryID = respd.Data.QueryID
+				rows.status = QueryStatusInProgress
+				rows.statusChannel = make(chan queryStatus)
+				respd.Data.AsyncRows = rows
+			default:
```
Comment (sfc-gh-anezvigin) [inline restful.go:224]: Should this default case fail query execution? The queryID, status, and statusChannel not being set feels like undefined behavior once the goroutine is kicked off below.

File: restful.go [hunk 7]
```diff
@@ -433,3 +461,69 @@ func cancelQuery(ctx context.Context, sr *snowflakeRestful, requestID string, ti
...
+		return
+	}
+
+	respd := execResponse{} // reset the response
+	err = json.NewDecoder(resp.Body).Decode(&respd)
+	resp.Body.Close()
+	if err != nil {
+		logger.WithContext(ctx).Errorf("failed to decode JSON. err: %v", err)
+		return
+	}
```
Comment (sfc-gh-anezvigin) [inline restful.go:487]: What's actually going to happen when this error is handled? Will rows.Next() block indefinitely?

You may wanna update statusChannel to not just receive a `queryStatus string` but a struct like:

```
type queryEvent struct {
  status queryStatus
  err error
}
```

Then update all of the checkAsyncExecStatus() functions to this signature:

```
checkAsyncExecStatus() error
```

Then update places like Next() to do something like this:

```
func (rows *snowflakeRows) Next(dest []driver.Value) (err error) {
    if err := rows.checkAsyncQueryStatus(); err != nil {
        return err
    }
```

That's just one idea though. The error handling flow would need to be tested (manually and I suggest writing unit tests for it as well).

File: statement_test.go [hunk 8]
```diff
@@ -497,3 +498,43 @@ func TestGetQueryID(t *testing.T) {
 		t.Fatalf("failed to prepare statement. err: %v", err)
 	}
 }
+
+func TestAsyncMode(t *testing.T) {
```
Comment (sfc-gh-anezvigin) [inline statement_test.go:502]: What does code coverage look like? Feels like there aren't enough test cases.

**Review by sfc-gh-anezvigin — CHANGES_REQUESTED** [review]
Overall approach looks good. Left some notes.

File: restful.go (hunk 7 above)
Comment (sfc-gh-jbahk) [reply restful.go:487]: This makes sense. I'll update it as such.

File: restful.go (hunk 6 above)
Comment (sfc-gh-jbahk) [reply restful.go:224]: Good point. I'll short circuit this and return the returned &respd object before the goroutine is kicked off.

File: statement_test.go (hunk 8 above)
Comment (sfc-gh-jbahk) [reply statement_test.go:502]: It's at 80.7% with the push I just made.

File: statement_test.go [hunk 9]
```diff
@@ -497,3 +498,91 @@ func TestGetQueryID(t *testing.T) {
...
```
Comment (sfc-gh-cshi) [inline statement_test.go:502]: I would suggest to add test cases in which several queries within the same session or in different sessions will be filed asynchronously, and some results of late queries could return earlier.

File: connection.go [hunk 10]
```diff
@@ -73,9 +69,18 @@ func (sc *snowflakeConn) exec(
...
 		SequenceID: counter,
 	}
-	req.IsInternal = isInternal
+	params := []paramKey{MultiStatementCount}
+	for _, k := range params {
+		key := ctx.Value(k)
+		if key != nil {
+			req.Parameters = map[string]interface{}{string(MultiStatementCount): key}
+		}
+	}
```
Comment (sfc-gh-anezvigin) [inline connection.go:81]: I can see what the for loop is trying to do (save on some typing later via the params slice), but this doesn't seem necessary. The following is much clearer albeit more boilerplate-y:

```
req := execRequest{
  // ...
  Parameters: map[string]interface{}{},
  // ...
}

if key := ctx.Value(MultiStatementCount); key != nil {
  req.Parameters[string(MultiStatementCount)] = key
}
if key := ctx.Value(SomeOtherThing); key != nil {
  req.Parameters[string(SomeOtherThing)] = key
}
if key := ctx.Value(YetAnotherThing); key != nil {
  req.Parameters[string(YetAnotherThing)] = key
}
```

File: connection.go [hunk 11]
```diff
@@ -547,6 +570,155 @@ func (sc *snowflakeConn) getQueryResult(ctx context.Context, resultPath string)
...
+			rows.statusChannel <- queryEvent{
+				QueryFailed,
+				&SnowflakeError{
+					Number:  -1,
+					Message: err.Error(),
+					QueryID: rows.queryID,
+				},
+			}
+			close(rows.statusChannel)
+		}
```
Comment (sfc-gh-anezvigin) [inline connection.go:619]: Can be DRY'd up significantly. The only thing different between both branches is the query ID and the channel. It's a non-trivial point because someone might update the error message for one SnowflakeError{} definition but forget the other one.

Comment (sfc-gh-anezvigin) [reply connection.go:619]: Actually what you could do is just define the channel at the very top of the function like:

```
var statusChannel chan queryEvent
if resType == "exec" {
  statusChannel = ...
} else {
  statusChannel = ...
}
```

This way you can do away with a lot of the if statements with duplicate logic in them.

File: connection.go [hunk 12]
```diff
@@ -547,6 +570,155 @@ func (sc *snowflakeConn) getQueryResult(ctx context.Context, resultPath string)
...
+		return
+	}
+
+	respd := execResponse{}
+	err = json.NewDecoder(resp.Body).Decode(&respd)
+	resp.Body.Close()
```
Comment (sfc-gh-anezvigin) [inline connection.go:625]: There's a small risk of a memory leak here, but I think it can be easily mitigated this way after calling FuncGet() above:

```
if resp.Body != nil {
  defer resp.Body.Close()
}
```

I _think_ (please verify) that it's okay to call resp.Body.Close() multiple times so you can still leave this close call in here.

The memory leak risk is there's a lot of code between this close call and the response body being created. If anything panics along the way (recovered later by the user) or if FuncGet() doesn't guarantee that `err != nil && resp = nil` -- you might have a leak that's hard to track down.

File: connection.go [hunk 13]
```diff
@@ -547,6 +570,155 @@ func (sc *snowflakeConn) getQueryResult(ctx context.Context, resultPath string)
...
+	if respd.Code != "" {
+		code, err = strconv.Atoi(respd.Code)
+		if err != nil {
+			code = -1
+		}
+	} else {
+		code = -1
+	}
+
+	if resType, _ := ctx.Value(resultType).(string); resType == "exec" {
```
Comment (sfc-gh-anezvigin) [inline connection.go:662]: Do we need to do this casting every time? Can it be done at the top of the func?

File: connection.go [hunk 14]
```diff
@@ -547,6 +570,155 @@ func (sc *snowflakeConn) getQueryResult(ctx context.Context, resultPath string)
...
+				&SnowflakeError{
+					Number:   code,
+					SQLState: respd.Data.SQLState,
+					Message:  respd.Message,
+					QueryID:  respd.Data.QueryID,
+				},
+			}
+		}
+		close(rows.statusChannel)
+	}
```
Comment (sfc-gh-anezvigin) [inline connection.go:719]: What is the else case here? You might as well panic so that channels don't block indefinitely. BTW if `restType` mapped to a constant of some kind it would decrease the likelihood of setting the wrong thing by accident. There are parts of the code that are doing this:

```
ctx = context.WithValue(ctx, resultType, "exec")
```

When it could be doing this:

```
ctx = setQueryResultType(ctx, ExecResultType)
```

With a complementary:

```
getQueryResultType(ctx) // ExecResultType | RowsResultType
```

File: result.go [hunk 15]
```diff
@@ -1,26 +1,76 @@
...
+		event := <-res.statusChannel
+		res.status = event.status
+		if event.err != nil {
+			res.err = event.err
+			return res.err
+		}
+		res.status = event.status
+	} else if res.status == QueryFailed {
+		return res.err
+	}
```
Comment (sfc-gh-anezvigin) [inline result.go:74]: Looking at this function made me realize... why not just make `res.statusChannel` more like `res.errorChannel` and have it be `chan error`? The only two things we send on this channel are complete events and errors. So if we receive `nil` on the channel that just means complete. Doing it this way minimizes the amount of states we need to worry about.

File: statement_test.go (hunk 8 above)
Comment (sfc-gh-anezvigin) [reply statement_test.go:502]: Got it, thanks. The main thing I'm worried about is making sure the edge cases are covered in tests and coverage helps highlight those.

**Review by sfc-gh-anezvigin — COMMENTED** [review]
Left a few more comments.

File: result.go (hunk 15 above)
Comment (sfc-gh-jbahk) [reply result.go:74]: This makes perfect sense. Will do.

File: connection.go (hunk 12 above)
Comment (sfc-gh-jbahk) [reply connection.go:625]: I confirmed that it's okay to call `.Close()` on a ReadCloser.

File: connection.go [hunk 16]
```diff
@@ -232,8 +235,13 @@ func (sc *snowflakeConn) ExecContext(ctx context.Context, query string, args []d
 	if sc.rest == nil {
 		return nil, driver.ErrBadConn
 	}
-	// TODO: handle noResult and isInternal
-	data, err := sc.exec(ctx, query, false, false, args)
+	noResult, err := isAsyncMode(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// TODO handle isInternal
```
Comment (sfc-gh-anezvigin) [inline connection.go:242]: Aren't we handling isInternal already? Above?

**Review by sfc-gh-anezvigin — APPROVED** [review]

File: connection.go (hunk 16 above)
Comment (sfc-gh-jbahk) [reply connection.go:242]: The feature isn't in place yet -- we just pass in the isInternal value but it's false by default currently.

File: connection.go [hunk 17]
```diff
@@ -547,6 +581,116 @@ func (sc *snowflakeConn) getQueryResult(ctx context.Context, resultPath string)
...
+		sfError.QueryID = rows.queryID
+	}
+	defer close(errChannel)
+	headers[headerAuthorizationKey] = fmt.Sprintf(headerSnowflakeToken, sr.Token)
+	resp, err := sr.FuncGet(ctx, sr, URL, headers, timeout)
+	if err != nil {
+		logger.WithContext(ctx).Errorf("failed to get response. err: %v", err)
+		sfError.Message = err.Error()
+		errChannel <- sfError
+		close(errChannel)
```
Comment (Yifeng-Sigma) [inline connection.go:624]: channel might close twice here
